File: sdk/ruby/test/wire/attachments_test.rb
# frozen_string_literal: true

require_relative "wiremock_test_case"

class AttachmentsWireTest < WireMockTestCase
  def setup
    super

    @client = AgentMailer::Client.new(
      api_key: "<token>",
      base_url: WIREMOCK_BASE_URL
    )
  end

  def test_attachments_upload_with_wiremock
    test_id = "attachments.upload.0"

    @client.attachments.upload(
      inbox_id: "inboxId",
      filename: "filename",
      content_type: "contentType",
      size: 1,
      checksum_sha256: "checksumSha256",
      request_options: {
        additional_headers: {
          "X-Test-Id" => "attachments.upload.0"
        }
      }
    )

    verify_request_count(
      test_id: test_id,
      method: "POST",
      url_path: "/v1/inboxes/inboxId/attachments",
      query_params: nil,
      expected: 1
    )
  end

  def test_attachments_complete_with_wiremock
    test_id = "attachments.complete.0"

    @client.attachments.complete(
      inbox_id: "inboxId",
      attachment_id: "attachmentId",
      request_options: {
        additional_headers: {
          "X-Test-Id" => "attachments.complete.0"
        }
      }
    )

    verify_request_count(
      test_id: test_id,
      method: "POST",
      url_path: "/v1/inboxes/inboxId/attachments/attachmentId/complete",
      query_params: nil,
      expected: 1
    )
  end

  def test_attachments_download_with_wiremock
    test_id = "attachments.download.0"

    @client.attachments.download(
      inbox_id: "inboxId",
      attachment_id: "attachmentId",
      request_options: {
        additional_headers: {
          "X-Test-Id" => "attachments.download.0"
        }
      }
    )

    verify_request_count(
      test_id: test_id,
      method: "GET",
      url_path: "/v1/inboxes/inboxId/attachments/attachmentId/download",
      query_params: nil,
      expected: 1
    )
  end
end

File: sdk/rust/tests/attachments_test.rs
use agentmailer::prelude::*;

mod wire_test_utils;

#[tokio::test]
#[allow(unused_variables, unreachable_code)]
async fn test_attachments_upload_with_wiremock() {
    wire_test_utils::reset_wiremock_requests().await.unwrap();
    let wiremock_base_url = wire_test_utils::get_wiremock_base_url();

    let mut config = ClientConfig {
        token: Some("<token>".to_string()),
        ..Default::default()
    };
    config.base_url = wiremock_base_url.to_string();
    let client = AgentMailer::new(config).expect("Failed to build client");

    let result = client
        .attachments
        .upload(
            &"inboxId".to_string(),
            &AttachmentUploadCreate {
                filename: "filename".to_string(),
                content_type: "contentType".to_string(),
                size: 1,
                checksum_sha256: "checksumSha256".to_string(),
            },
            None,
        )
        .await;

    assert!(result.is_ok(), "Client method call should succeed");

    wire_test_utils::verify_request_count("POST", "/v1/inboxes/inboxId/attachments", None, 1)
        .await
        .unwrap();
}

#[tokio::test]
#[allow(unused_variables, unreachable_code)]
async fn test_attachments_complete_with_wiremock() {
    wire_test_utils::reset_wiremock_requests().await.unwrap();
    let wiremock_base_url = wire_test_utils::get_wiremock_base_url();

    let mut config = ClientConfig {
        token: Some("<token>".to_string()),
        ..Default::default()
    };
    config.base_url = wiremock_base_url.to_string();
    let client = AgentMailer::new(config).expect("Failed to build client");

    let result = client
        .attachments
        .complete(&"inboxId".to_string(), &"attachmentId".to_string(), None)
        .await;

    assert!(result.is_ok(), "Client method call should succeed");

    wire_test_utils::verify_request_count(
        "POST",
        "/v1/inboxes/inboxId/attachments/attachmentId/complete",
        None,
        1,
    )
    .await
    .unwrap();
}

#[tokio::test]
#[allow(unused_variables, unreachable_code)]
async fn test_attachments_download_with_wiremock() {
    wire_test_utils::reset_wiremock_requests().await.unwrap();
    let wiremock_base_url = wire_test_utils::get_wiremock_base_url();

    let mut config = ClientConfig {
        token: Some("<token>".to_string()),
        ..Default::default()
    };
    config.base_url = wiremock_base_url.to_string();
    let client = AgentMailer::new(config).expect("Failed to build client");

    let result = client
        .attachments
        .download(&"inboxId".to_string(), &"attachmentId".to_string(), None)
        .await;

    assert!(result.is_ok(), "Client method call should succeed");

    wire_test_utils::verify_request_count(
        "GET",
        "/v1/inboxes/inboxId/attachments/attachmentId/download",
        None,
        1,
    )
    .await
    .unwrap();
}
